mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala

```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
  * Can return weights in Vector format.
  */
 private[ann] trait LayerModel extends Serializable {
-  /**
-   * number of weights
-   */
-  val size: Int
 
+  val weights: BDV[Double]
   /**
    * Evaluates the data (process the data through the layer)
    * @param data data
-   * @return processed data
+   * @param output output to write to
    */
-  def eval(data: BDM[Double]): BDM[Double]
+  def eval(data: BDM[Double], output: BDM[Double]): Unit
 
   /**
    * Computes the delta for back propagation
-   * @param nextDelta delta of the next layer
-   * @param input input data
+   * @param delta delta of this layer
+   * @param output output of this layer
+   * @param pDelta storage for the result, the previous delta
```

Comment:
* document what input parameters will be changed in-place
* `prevDelta`. We can change the method name to `computePrevDelta` (using a verb)